File: frontend/src/lib/domain/facility/copy-job.ts
export type FacilityJobKind =
  | 'control_cabinet'
  | 'sps_controller'
  | 'sps_controller_system_type'
  | 'field_device'
  | 'object_data';
export type FacilityJobStatus = 'queued' | 'running' | 'completed' | 'failed';
export type FacilityJobType = 'copy' | 'export' | 'bulk' | 'delete' | 'restore';
export type FacilityJobClass = 'mutation' | 'export';

export interface FacilityJobResult {
  download_url?: string;
  file_name?: string;
  output_type?: 'excel' | 'zip';
  content_type?: string;
  size?: number;
  expires_at?: string;
  resource_id?: string;
  [key: string]: unknown;
}

export interface FacilityJob {
  jobId: string;
  kind: FacilityJobKind;
  type: FacilityJobType;
  class: FacilityJobClass;
  status: FacilityJobStatus;
  progress: number;
  stage: string;
  error?: string;
  attempts: number;
  processed: number;
  total?: number;
  successCount: number;
  failureCount: number;
  retryable: boolean;
  result?: FacilityJobResult;
  createdAt?: string;
  updatedAt?: string;
  completedAt?: string;
}

interface FacilityJobWire {
  job_id?: string;
  kind?: string;
  type?: string;
  class?: string;
  status?: string;
  progress?: number;
  stage?: string;
  error?: string;
  attempts?: number;
  processed?: number;
  total?: number;
  success_count?: number;
  failure_count?: number;
  retryable?: boolean;
  result?: FacilityJobResult;
  created_at?: string;
  updated_at?: string;
  completed_at?: string;
}

export function toFacilityJob(response: FacilityJobWire): FacilityJob {
  if (!response.job_id || !isKind(response.kind) || !isStatus(response.status) || !response.stage) {
    throw new Error('Invalid facility job response');
  }
  return {
    jobId: response.job_id,
    kind: response.kind,
    type: isType(response.type) ? response.type : 'copy',
    class: response.class === 'export' ? 'export' : 'mutation',
    status: response.status,
    progress: Math.min(100, Math.max(0, response.progress ?? 0)),
    stage: response.stage,
    attempts: response.attempts ?? 0,
    processed: response.processed ?? 0,
    ...(response.total !== undefined ? { total: response.total } : {}),
    successCount: response.success_count ?? 0,
    failureCount: response.failure_count ?? 0,
    retryable: response.retryable ?? false,
    ...(response.error ? { error: response.error } : {}),
    ...(response.result ? { result: response.result } : {}),
    ...(response.created_at ? { createdAt: response.created_at } : {}),
    ...(response.updated_at ? { updatedAt: response.updated_at } : {}),
    ...(response.completed_at ? { completedAt: response.completed_at } : {})
  };
}

function isKind(value: unknown): value is FacilityJobKind {
  return (
    value === 'control_cabinet' ||
    value === 'sps_controller' ||
    value === 'sps_controller_system_type' ||
    value === 'field_device' ||
    value === 'object_data'
  );
}

function isStatus(value: unknown): value is FacilityJobStatus {
  return value === 'queued' || value === 'running' || value === 'completed' || value === 'failed';
}

function isType(value: unknown): value is FacilityJobType {
  return (
    value === 'copy' ||
    value === 'export' ||
    value === 'bulk' ||
    value === 'delete' ||
    value === 'restore'
  );
}

export type CopyJob = FacilityJob;
export type CopyJobKind = FacilityJobKind;
export type CopyJobStatus = FacilityJobStatus;
export const toCopyJob = toFacilityJob;

File: backend/internal/handler/dto/facility/copy_job.go
package facility

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// CopyJobResponse is returned immediately after starting a hierarchy copy and
// is also used to reconcile a job after a WebSocket reconnect.
type CopyJobResponse struct {
	JobID        uuid.UUID       `json:"job_id"`
	Kind         string          `json:"kind" enums:"control_cabinet,sps_controller,sps_controller_system_type,field_device,object_data"`
	Type         string          `json:"type" enums:"copy,export,bulk,delete,restore"`
	Class        string          `json:"class" enums:"mutation,export"`
	Status       string          `json:"status" enums:"queued,running,completed,failed"`
	Progress     int             `json:"progress" minimum:"0" maximum:"100"`
	Stage        string          `json:"stage"`
	Error        string          `json:"error,omitempty"`
	Attempts     int             `json:"attempts"`
	Processed    int64           `json:"processed"`
	Total        *int64          `json:"total,omitempty"`
	SuccessCount int64           `json:"success_count"`
	FailureCount int64           `json:"failure_count"`
	Retryable    bool            `json:"retryable"`
	Result       json.RawMessage `json:"result,omitempty" swaggertype:"object"`
	CreatedAt    time.Time       `json:"created_at"`
	UpdatedAt    time.Time       `json:"updated_at"`
	CompletedAt  *time.Time      `json:"completed_at,omitempty"`
}

// FacilityJobResponse is the canonical name. CopyJobResponse remains as a
// compatibility contract for the existing copy-job routes.
type FacilityJobResponse = CopyJobResponse

type FacilityJobListResponse struct {
	Items          []FacilityJobResponse `json:"items"`
	NextCursor     string                `json:"next_cursor,omitempty"`
	PreviousCursor string                `json:"previous_cursor,omitempty"`
}
